# Matrix Inversion using Gauss elimination method

In [ ]:
from sympy import Matrix, Rational, latex
from IPython.display import display, Markdown, Math, HTML

class InvertibleMatrix:
    def __init__(self, matrix):
        """
        Initializes a matrix to be inverted using the Gauss-Jordan method.

        Parameters:
        - matrix: The square matrix to be inverted.
        """
        # Convert all entries to Rational numbers
        self.matrix = Matrix(matrix).applyfunc(Rational)
        self.operations = []

        # Check if the matrix is square
        if self.matrix.rows != self.matrix.cols:
            raise ValueError("The matrix must be square.")

        # Create the augmented matrix with the identity matrix (with Rational entries)
        identity = Matrix.eye(self.matrix.rows).applyfunc(Rational)
        self.aug_matrix = self.matrix.row_join(identity)

        display(Markdown("**Initial Matrix (Starting matrix):**"))
        self.display_matrix()

    def __repr__(self):
        return repr(self.aug_matrix)

    def __str__(self):
        return str(self.aug_matrix)

    def _repr_latex_(self):
        return self.aug_matrix._repr_latex_()

    def _validate_row_number(self, row):
        if not isinstance(row, int):
            raise TypeError("Row number must be an integer.")
        if row < 1 or row > self.aug_matrix.rows:
            raise IndexError(f"Row number must be in the range from 1 to {self.aug_matrix.rows}.")
        return row - 1

    def add_row(self, target_row, source_row, coefficient):
        target_idx = self._validate_row_number(target_row)
        source_idx = self._validate_row_number(source_row)
        coefficient = Rational(coefficient)

        self.aug_matrix.row_op(target_idx, lambda v, j: v + coefficient * self.aug_matrix[source_idx, j])

        operation_str = f"r{target_row} = r{target_row} + {coefficient}*r{source_row}"
        self.operations.append(operation_str)
        display(Markdown(f"**Operation:** {operation_str}"))
        self.display_matrix()

    def multiply_row(self, row, coefficient):
        row_idx = self._validate_row_number(row)
        coefficient = Rational(coefficient)

        self.aug_matrix.row_op(row_idx, lambda v, _: coefficient * v)

        operation_str = f"r{row} = {coefficient}*r{row}"
        self.operations.append(operation_str)
        display(Markdown(f"**Operation:** {operation_str}"))
        self.display_matrix()

    def swap_rows(self, row1, row2):
        row1_idx = self._validate_row_number(row1)
        row2_idx = self._validate_row_number(row2)

        self.aug_matrix.row_swap(row1_idx, row2_idx)

        operation_str = f"Swap r{row1} <-> r{row2}"
        self.operations.append(operation_str)
        display(Markdown(f"**Operation:** {operation_str}"))
        self.display_matrix()

    def display_matrix(self):
        """Displays the left and right matrix side by side in LaTeX format."""
        left_matrix = self.aug_matrix[:, :self.matrix.cols]
        right_matrix = self.aug_matrix[:, self.matrix.cols:]

        # Generate LaTeX code for both matrices
        left_latex = latex(left_matrix)
        right_latex = latex(right_matrix)

        # Combine both matrices into a single display output
        combined_latex = r"""
        %s
        \quad
        %s
        """ % (left_latex, right_latex)

        display(Math(combined_latex))

    def print_operations(self):
        display(Markdown("**Performed Operations:**"))
        for op in self.operations:
            print(op)

    def get_inverse(self):
        """Returns the inverse of the matrix after performing Gauss-Jordan elimination."""
        # Check if the left part of the augmented matrix is the identity matrix
        left_matrix = self.aug_matrix[:, :self.matrix.cols]
        if not left_matrix == Matrix.eye(self.matrix.rows):
            raise ValueError("The matrix has not been reduced to the identity matrix. Continue the operations.")
        # Return the right part of the augmented matrix as the inverse
        inverse_matrix = self.aug_matrix[:, self.matrix.cols:]
        display(Markdown("**Inverse Matrix:**"))
        display(Math(latex(inverse_matrix)))
        return inverse_matrix

**Example 1:**

In [6]:
# Create an instance of the class with a matrix to be inverted
initial_matrix = [[2, 1], [5, 3]] # 2x2 matrix
m = InvertibleMatrix(initial_matrix) # Create an instance of the class

**Initial Matrix (Starting matrix):**

<IPython.core.display.Math object>

In [11]:
import sympy as sp # import the sympy library
a = sp.Matrix(initial_matrix) # create the initial matrix
print("The inverse matrix is:")
a.inv() # calculate the inverse matrix

The inverse matrix is:


Matrix([
[7/4, -3/8, -5/8],
[ -3,    1,    1],
[1/4, -1/8,  1/8]])

In [ ]:
# Add -5/2 times "row 1" to "row 2"
m.add_row(2, 1, -5/2)

**Operation:** r2 = r2 + -5/2*r1

<IPython.core.display.Math object>

In [ ]:
# Multiply "row 1" by 1/2
m.multiply_row(1, 1/2)

**Operation:** r1 = 1/2*r1

<IPython.core.display.Math object>

In [ ]:
# Add -1 times "row 2" to "row 1"
m.add_row(1, 2, -1)

**Operation:** r1 = r1 + -1*r2

<IPython.core.display.Math object>

In [ ]:
# Multiply "row 2" by 2
m.multiply_row(2, 2)

**Operation:** r2 = 2*r2

<IPython.core.display.Math object>

The matrix has been correctly computed!

**Example 2**

In [7]:
initial_matrix = [[2, 1, 2], [5, 3, 1], [1, 1, 5]] # 3x3 matrix
m = InvertibleMatrix(initial_matrix) # Create an instance of the class
sympy_m = sp.Matrix(initial_matrix) # create the initial matrix

**Initial Matrix (Starting matrix):**

<IPython.core.display.Math object>

In [ ]:
inverse = sympy_m.inv() # calculate the inverse matrix
inverse

Matrix([
[7/4, -3/8, -5/8],
[ -3,    1,    1],
[1/4, -1/8,  1/8]])

In [10]:
m.add_row(2, 3, -5)

**Operation:** r2 = r2 + -5*r3

<IPython.core.display.Math object>

In [ ]:
m.add_row(3, 1, -1/2)

**Operation:** r3 = r3 + -1/2*r1

<IPython.core.display.Math object>

In [ ]:
m.multiply_row(3, 4)

**Operation:** r3 = 4*r3

<IPython.core.display.Math object>

In [ ]:
m.add_row(3, 2, 1)

**Operation:** r3 = r3 + 1*r2

<IPython.core.display.Math object>

In [ ]:
m.multiply_row(3, -1/8)

**Operation:** r3 = -1/8*r3

<IPython.core.display.Math object>

In [ ]:
m.multiply_row(2, -1/2)

**Operation:** r2 = -1/2*r2

<IPython.core.display.Math object>

In [ ]:
m.add_row(2, 3, -12)

**Operation:** r2 = r2 + -12*r3

<IPython.core.display.Math object>

In [ ]:
m.add_row(1, 3, -2)

**Operation:** r1 = r1 + -2*r3

<IPython.core.display.Math object>

In [ ]:
m.add_row(1, 2, -1)

**Operation:** r1 = r1 + -1*r2

<IPython.core.display.Math object>

In [ ]:
m.multiply_row(1, 1/2)

**Operation:** r1 = 1/2*r1

<IPython.core.display.Math object>

---

## Exercises for Students

Find the inverse matrices using the Gauss method:

$$
A=
\begin{bmatrix}
1 & 2\\
3 & 4
\end{bmatrix}
, \qquad
B=
\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 1 \\
2 & 3 & 2
\end{bmatrix}
,\qquad
C=
\begin{bmatrix}
0 & 0 & 1\\
0 & 1 & 0\\
1 & 0 & 0
\end{bmatrix}
$$

In [18]:
import numpy as np

# Define the matrices
A = np.array([[1, 2], [3, 4]])
B = np.array([[1, 2, 3], [4, 5, 1], [2, 3, 2]])
C = np.array([[0, 0, 1], [0, 1, 0], [1, 0, 0]])

# Function to compute the inverse of a matrix using numpy
def inverse_matrix(matrix):
    try:
        inv_matrix = np.linalg.inv(matrix)
        return inv_matrix
    except np.linalg.LinAlgError:
        return "Matrix is singular and cannot be inverted."

# Compute inverses
inverse_A = inverse_matrix(A)
inverse_B = inverse_matrix(B)
inverse_C = inverse_matrix(C)

inverse_A, inverse_B, inverse_C


(array([[-2. ,  1. ],
        [ 1.5, -0.5]]),
 array([[  7.,   5., -13.],
        [ -6.,  -4.,  11.],
        [  2.,   1.,  -3.]]),
 array([[0., 0., 1.],
        [0., 1., 0.],
        [1., 0., 0.]]))

### **Finding Inverse Matrices by Hand Using the Gauss Method**

---

### **1. Find the Inverse of Matrix \( A \):**

Matrix \( A = \begin{bmatrix} 1 & 2 \\ 3 & 4 \end{bmatrix} \).

#### Steps:

1. **Write the augmented matrix:**
   \
   \begin{array}{cc|cc}
   1 & 2 & 1 & 0 \\
   3 & 4 & 0 & 1
   \end{array}
   

2. **Make the first element of the first row \( 1 \) (it already is).**

3. **Zero out the element below the first pivot (in the second row):**
   - Subtract \( 3 \times \text{(first row)} \) from the second row:
   \begin{array}{cc|cc}
   1 & 2 & 1 & 0 \\
   0 & -2 & -3 & 1
   \end{array}

4. **Make the pivot in the second row (second column) equal to \( 1 \):**
   - Divide the second row by \( -2 \):
  \begin{array}{cc|cc}
   1 & 2 & 1 & 0 \\
   0 & 1 & 1.5 & -0.5
   \end{array}

5. **Zero out the element above the second pivot (in the first row):**
   - Subtract \( 2 \times \text{(second row)} \) from the first row:
  \begin{array}{cc|cc}
   1 & 0 & -2 & 1 \\
   0 & 1 & 1.5 & -0.5
   \end{array}

6. **Extract the inverse matrix:**
   \[
   A^{-1} = \begin{bmatrix}
   -2 & 1 \\
   1.5 & -0.5
   \end{bmatrix}


---

### **2. Find the Inverse of Matrix \( B \):**

Matrix \( B = \begin{bmatrix} 1 & 2 & 3 \\ 4 & 5 & 1 \\ 2 & 3 & 2 \end{bmatrix} \).

#### Steps:

1. **Write the augmented matrix:**
   [\begin{array}{ccc|ccc}
   1 & 2 & 3 & 1 & 0 & 0 \\
   4 & 5 & 1 & 0 & 1 & 0 \\
   2 & 3 & 2 & 0 & 0 & 1
   \end{array}

2. **Make the first pivot \( 1 \) (it already is).**

3. **Zero out the elements below the first pivot (second and third rows):**
   - Subtract \( 4 \times \text{(first row)} \) from the second row.
   - Subtract \( 2 \times \text{(first row)} \) from the third row.

4. **Make the second pivot \( 1 \):**
   - Divide the second row by the value of the pivot.

5. **Zero out the elements above and below the second pivot:**
   - Use the second row to adjust the first and third rows.

6. **Make the third pivot \( 1 \):**
   - Divide the third row by the value of the pivot.

7. **Zero out the elements above the third pivot:**
   - Use the third row to adjust the first and second rows.

8. **Extract the inverse matrix.**

---

### **3. Find the Inverse of Matrix \( C \):**

Matrix \( C = \begin{bmatrix} 0 & 0 & 1 \\ 0 & 1 & 0 \\ 1 & 0 & 0 \end{bmatrix} \).

#### Steps:

1. **Recognize that this is a permutation matrix.**
   - The inverse of a permutation matrix is its transpose.

2. **Transpose \( C \):**
   
   C^{-1} = \begin{bmatrix}
   0 & 0 & 1 \\
   0 & 1 & 0 \\
   1 & 0 & 0
   \end{bmatrix}
  

---

### **Summary:**

For each matrix, the key is to apply elementary row operations to transform the original matrix into the identity matrix while applying the same operations to the identity matrix to compute the inverse.


